libvast/native-plugins/compute.cpp

```diff
@@ -0,0 +1,152 @@
+    transformed_batches_.emplace_back(std::move(layout), std::move(batch));
+    return caf::none;
+  }
+
+  [[nodiscard]] caf::expected<std::vector<transform_batch>> finish() override {
+    return std::exchange(transformed_batches_, {});
+  }
+
+private:
+  /// Cache for transformed batches.
+  std::vector<transform_batch> transformed_batches_ = {};
+
+  /// Cache for the per-layout transformations.
+  std::unordered_map<type, std::vector<indexed_transformation>>
+    indexed_transformations_cache_ = {};
+
+  /// The paramters of the transformation.
+  std::shared_ptr<arrow::compute::Function> function_ = {};
+  std::unique_ptr<arrow::compute::FunctionOptions> function_options_ = {};
+  std::vector<std::string> fields_ = {};
+  std::vector<arrow::Datum> inputs_ = {};
+};
+
+// -- plugin ------------------------------------------------------------------
+
+class plugin final : public virtual transform_plugin {
+public:
+  caf::error initialize(data options) override {
+    // We don't use any plugin-specific configuration under
+    // vast.plugins.compute, so nothing is needed here.
+    if (caf::holds_alternative<caf::none_t>(options))
+      return caf::none;
+    if (const auto* rec = caf::get_if<record>(&options))
+      if (rec->empty())
+        return caf::none;
+    return caf::make_error(ec::invalid_configuration, "expected empty "
+                                                      "configuration under "
+                                                      "vast.plugins.compute");
+  }
+
+  [[nodiscard]] const char* name() const override {
+    return "compute";
+  };
+
+  [[nodiscard]] caf::expected<std::unique_ptr<transform_step>>
+  make_transform_step(const record&) const override {
+    // FIXME: parse function name, function options, fields, and inputs from
+    // configuration.
+    const auto* registry = arrow::compute::GetFunctionRegistry();
+    auto get_function_result = registry->GetFunction("ascii_trim");
+    if (!get_function_result.ok())
+      return caf::make_error(ec::unspecified,
+                             get_function_result.status().ToString());
+    auto function = get_function_result.MoveValueUnsafe();
+    if (function->kind() != arrow::compute::Function::SCALAR)
+      return caf::make_error(ec::unimplemented, "non-scalar compute function");
+    // FIXME: support non-unary functions.
+    if (function->arity().num_args != 1)
+      return caf::make_error(ec::unimplemented, "non-unary compute function");
+    auto function_options = std::make_unique<arrow::compute::TrimOptions>(". ");
+    auto fields = std::vector<std::string>{"hostname"};
+    auto inputs = std::vector<arrow::Datum>{};
+    return std::make_unique<compute_step>(std::move(function),
+                                          std::move(function_options),
+                                          std::move(fields), std::move(inputs));
+    return ec::unimplemented;
+  }
+};
```

Comment:
@mavam I couldn't quite let this idea go: generally speaking, we "only" need to find some way to create `function_options` and `inputs` from the YAML configuration. Everything else is trivial.

This hardcoded version actually works nicely for the use case that was originally proposed by @satta as a trim transform step with the given example configuration.

Reply:
> generally speaking, we "only" need to find some way to create function_options and inputs from the YAML configuration. Everything else is trivial.

Yep, that was my understanding of the crux as well.

Reply:
My current understanding is that we've just have to map them all by hand. It's not that many, though, and we gain the advantage of helpful error messages if we parse ourselves.

Here's a list of the possible options: https://arrow.apache.org/docs/cpp/api/compute.html#concrete-options-classes